Feature/check iap eligibility

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->
Part of: #7925
<!-- Id number of the GitHub issue this PR addresses. -->

### Description

First step of IAP integration in store creation flow: check if user is eligible to use IAP. 

- If user is eligible, the native flow will be triggered 
- If user is not eligible, either because of an error while checking eligibility or because IAP is not supported for the Google account on the device a dialog will be shown explaining store creation flow is not available in their country. 

The feature is currently hidden under a feature flag and won't be enabled unless we are rejected from one of the app stores.  

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Follow these instructions: PdfdoF-1Ej-p2 (Android section) to get an eligible account for IAP and test the feature. 
- Check that triggering store creation with an eligible account in `debug` app navigates to native store creation flow
- Check that triggering store creation in `debug` app with a non IAP eligible account displays the error dialog
- Check that if feature flag `IAP_FOR_STORE_CREATION -> false` then the native store creation flow is opened as usual

### Images/gif

Success 


https://user-images.githubusercontent.com/2663464/208727825-a08ed355-1860-4ec6-8f87-ebe5150697ac.mp4


Error

https://user-images.githubusercontent.com/2663464/208727058-feff864a-4ed8-45e3-9749-11a64a18b08e.mp4





- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

JorgeMucientes committed:
Add provider to inject PurchaseWPComPlanActions

JorgeMucientes committed:
Add new screen to load iap user eligibility

JorgeMucientes committed:
Check IAP eligibility when opening native store creation flow

JorgeMucientes committed:
Handle navigation after checking IAP eligibility

JorgeMucientes committed:
Merge branch 'trunk' into feature/check-iap-eligibility

JorgeMucientes committed:
Use new planSupportChecker to check user eligibility

JorgeMucientes committed:
Add correct navigation event

JorgeMucientes committed:
Add tracking after IAP eligibility is fetched

JorgeMucientes committed:
Fix indentation issues

JorgeMucientes committed:
Merge branch 'trunk' into feature/check-iap-eligibility

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/analytics/AnalyticsEvent.kt

JorgeMucientes committed:
Fix race condition when checking IAP

JorgeMucientes committed:
Add feature flag for using IAP on store creation

JorgeMucientes committed:
Rename event to be consistent with the rest of the events in the flow

JorgeMucientes committed:
Merge branch 'trunk' into feature/check-iap-eligibility

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/iapeligibility/IapEligibilityViewModel.kt:
```diff
@@ -0,0 +1,61 @@
+    savedStateHandle: SavedStateHandle,
+    private val planSupportChecker: PurchaseWpComPlanSupportChecker,
+    private val analyticsTrackerWrapper: AnalyticsTrackerWrapper
+) : ScopedViewModel(savedStateHandle) {
```
`planSupportChecker` will be passed as parameter of the ViewModel constructor in subsequent PR [here](https://github.com/woocommerce/woocommerce-android/pull/8053/files#diff-dfa5f27a6b0081652f06322fb8f3a2a56de2e3e3cc21cb6a83d24c56858f134e).

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/di/InAppPurchasesModule.kt:
```diff
@@ -0,0 +1,35 @@
+import dagger.hilt.components.SingletonComponent
+
+@InstallIn(SingletonComponent::class)
+@Module
```
Changes to this hilt module have been applied [here](https://github.com/woocommerce/woocommerce-android/pull/8053/files#diff-7c0f991eb187df783e5bb1441a22502f8b59a842d712c92e989bf961ddddcc56) to scope the IAP classes to the view model life cycle

JorgeMucientes committed:
Handle error cases and when IAP is not supported for user's country

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/di/InAppPurchasesModule.kt:
```diff
@@ -0,0 +1,35 @@
+import com.woocommerce.android.iap.pub.IAPSitePurchasePlanFactory
+import com.woocommerce.android.iap.pub.PurchaseWPComPlanActions
+import com.woocommerce.android.iap.pub.PurchaseWpComPlanSupportChecker
+import com.woocommerce.android.iapshowcase.IAPDebugLogWrapper
```
`iapshowcase` package is available only in debug, so you can not use it in production code. Probably you need to have your own log wrapper and IAPShowcaseMobilePayAPIProvider

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/iapeligibility/CheckIapEligibilityFragment.kt:
```diff
@@ -0,0 +1,104 @@
+                is NavigateToWebStoreCreation -> navigateToStoreCreationWeb()
+                is NavigateToNextStep -> navigateToStoreCreationNative()
+                is Exit -> findNavController().popBackStack()
+                is MultiLiveEvent.Event.ShowDialog -> event.showDialog()
```
np: would be good to hide `CircularProgressIndicator` when the dialog is shown. Now it is visible behind the dialog and might be a little bit confusing for a user

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/iapeligibility/IapEligibilityViewModel.kt:
```diff
@@ -0,0 +1,74 @@
+import javax.inject.Inject
+
+@HiltViewModel
+class IapEligibilityViewModel @Inject constructor(
```
Maybe a unit test on this class?

JorgeMucientes replied:
Yes! I will add it in [this PR ](https://github.com/woocommerce/woocommerce-android/pull/8053) so I can have the latest changes from branch `feature/iap-payment-flow`

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/iapeligibility/IapEligibilityViewModel.kt:
```diff
@@ -0,0 +1,74 @@
+                }
+            }
+        } else {
+            triggerEvent(NavigateToNextStep)
```
Shell we go to the web here?

JorgeMucientes replied:
Not really. If IAP feature flag is disabled we keep using the native flow only that at the moment of purchasing/checking out the eCommerce plan, the user will be redirected to a "web view checkout flow", but 90% of the flow will still be native. 
That is also the reason why I called the event `NavigateToNextStep`. Because that is consistent with the naming of all the navigating event along the store creation native flow.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/iapeligibility/IapEligibilityViewModel.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    sealed class IapEligibilityEvent : MultiLiveEvent.Event() {
+        object NavigateToNextStep : IapEligibilityEvent()
```
np: maybe call it `NavigateToNativeStep` to be more explicit?

JorgeMucientes replied:
As explained in the previous comment, the store creation web flow doesn't exist anymore, so the "next step" will always be native. I don't see any reason to be explicit about it. But good point 👍🏼

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/analytics/AnalyticsTracker.kt:
```diff
@@ -425,6 +425,7 @@ class AnalyticsTracker private constructor(private val context: Context) {
         const val KEY_PATH = "path"
 
         // -- Store creation
+        const val KEY_IAP_ELIGIBLE = "isEligible"
```
Just noticed in the logs:

```
2022-12-21 18:02:40.625 20529-20529 NosaraEvent             com.woocommerce.android              W  Properties should have lowercase name: isEligible
2022-12-21 18:02:40.626 20529-20529 NosaraEvent             com.woocommerce.android              E  Cannot add the property: isEligible to the event. Property name must match: ^[a-z_][a-z0-9_]*$
```

JorgeMucientes committed:
Add WooIapLogWrapper to be used in iap processing

JorgeMucientes committed:
Fix tracking key format

kidinov reviewed (changes requested):
Overal looks good to me! Thanks!

 I left some comments + CI should be fixed

JorgeMucientes committed:
Hide loading state when IAP eligibility error shown

JorgeMucientes committed:
Update message on the screen in case dialog is dismissed

JorgeMucientes committed:
Merge branch 'trunk' into feature/check-iap-eligibility

# Conflicts:
#	WooCommerce/src/main/res/values/strings.xml

JorgeMucientes commented:
Hey @kidinov thanks for the great feedback! I've addressed all the suggestions. If now I'll leave a comment explaining why not. 

Ready for another round :)

JorgeMucientes committed:
Merge branch 'trunk' into feature/check-iap-eligibility

kidinov reviewed (approved):
LGTM to me!
